runtime/sema/check_member_expression.go

```diff
@@ -443,12 +422,69 @@ func (checker *Checker) isReadableMember(accessedType Type, member *Member, acce
 			return true, member.Access
 		}
 	case EntitlementMapAccess:
-		return mapAccess(access, accessedType)
+		return checker.mapAccess(access, accessedType, resultingType, accessRange)
 	}
 
 	return false, member.Access
 }
 
+func (checker *Checker) mapAccess(
+	mappedAccess EntitlementMapAccess,
+	accessedType Type,
+	resultingType Type,
+	accessRange func() ast.Range,
+) (bool, Access) {
+
+	switch ty := accessedType.(type) {
+	case *ReferenceType:
+		// when accessing a member on a reference, the read is allowed, but the
+		// granted entitlements are based on the image through the map of the reference's entitlements
+		grantedAccess, err := mappedAccess.Image(ty.Authorization, accessRange)
+		if err != nil {
+			checker.report(err)
+			return false, mappedAccess
+		}
+		return true, grantedAccess
+
+	case *OptionalType:
+		return checker.mapAccess(mappedAccess, ty.Type, resultingType, accessRange)
+
+	default:
+		if mappedAccess.Type == IdentityMappingType {
+			access := AllSupportedEntitlements(resultingType)
+			if access != nil {
+				return true, access
+			}
+		}
+
+		// when accessing a member on a non-reference, the resulting mapped entitlement
+		// should be the entire codomain of the map
+		return true, mappedAccess.Codomain()
+	}
+}
+
+func AllSupportedEntitlements(typ Type) Access {
+	switch typ := typ.(type) {
+	case *ReferenceType:
+		return AllSupportedEntitlements(typ.Type)
+	case *OptionalType:
+		return AllSupportedEntitlements(typ.Type)
+	case *FunctionType:
+		return AllSupportedEntitlements(typ.ReturnTypeAnnotation.Type)
```

Comment:
This is used for accessor functions right? That use case makes sense, but we should add a test to make sure that a function-typed field with an identity mapping should not get the entitlements to return type. I.e. some field defined with `access(Identity) let fn: ((): T)` should not be entitled to `T` when it's accessed through an owned value.

Reply:
Didn't quite understand the question. For an owned value, `fn` won't return a reference, so we are not actually computing any entitlements for reference, the field is readable for anyone since the value is owned, correct?

On the other hand, it is not allowed to use Mappings on function-typed fields. Added a test e2ccbb0edf51eee1f7af6a8f520fa31b4875252d.

Reply:
Sorry, I should have been clearer. In https://github.com/onflow/cadence/pull/2637#discussion_r1258458218 I suggested adding a field with a `&((): T)` type to exercise this case.

Reply:
I see what you mean.

Weird enough, in stable cadence (not just this feature branch), accessing a field of type `&((): T)`, returns a `((): T)`. Don't know if this was changed intentionally, or was a regression (to me it seems like a regression, and probably needs fixing). So the use-case you mentioned, is currently not possible. But if we change that other behavior, then yeah, we might run into what you said.

I will add a fix to that hypothetical scenario for now, so it should seamlessly work, once the other issue is fixed.

Good catch nevertheless!

Reply:
added in 842a9cd482a53e47943756ca6817c45774c531ea